col/src/main/java/vct/col/ast/util/NameScanner.scala

```diff
@@ -0,0 +1,210 @@
+package vct.col.ast.util
+
+import java.util
+import java.util.Objects
+import scala.annotation.varargs
+import scala.collection.mutable
+import scala.collection.JavaConverters._
+import vct.col.ast.expr._
+import vct.col.ast.stmt.composite.BlockStatement
+import vct.col.ast.stmt.decl.DeclarationStatement
+import vct.col.ast.stmt.composite.ForEachLoop
+import vct.col.ast.stmt.composite.LoopStatement
+import vct.col.ast.stmt.composite.ParallelBlock
+import vct.col.ast.`type`.Type
+import NameExpressionKind._
+import vct.col.ast.generic.ASTNode
+import vct.col.ast.stmt.composite.VectorBlock
+import vct.col.ast.stmt.terminal.AssignmentStatement
+
+object NameScanner {
+  def freeVars[R <: ASTNode](nodes: util.List[R]): util.Map[String, Type] = {
+    val scanner = new NameScanner
+    for (n <- nodes.asScala) {
+      n.accept(scanner)
+    }
+    scanner.freeNamesJava
+  }
+
+  @varargs
+  def freeVars(theNodes: ASTNode*): util.Map[String, Type] = freeVars(theNodes.asJava)
+}
+
+class NameScanner extends RecursiveVisitor[AnyRef](null, null) {
+
+  case class Entry(name: String, typ: Type, var writtenTo: Boolean)
+
+  val frameStack: mutable.Stack[mutable.Set[DeclarationStatement]] = mutable.Stack()
```

Comment:
Scala docs: "@deprecated (Since version 2.12.0) Stack is an inelegant and potentially poorly-performing wrapper around List. Use a List assigned to a var instead."

Reply:
In 2.13.0 it is not deprecated anymore, and the interface fits, so I would like to keep using it. Couldn't find a "@ignore-deprecated-warning" annotation or something similar.